air-script/tests/aux_trace/aux_trace.masm

```diff
@@ -105,8 +135,25 @@ proc.evaluate_integrity_constraints
 end # END PROC evaluate_integrity_constraints
 
 proc.evaluate_boundary_constraints
-    exec.compute_evaluate_boundary_constraints
-    # Accumulate the numerator of the constraint polynomial
-    ext2add ext2add ext2add ext2add ext2add ext2add
+    exec.compute_boundary_constraints_aux_last
+    # Accumulate the numerator for segment 1 LastRow
+    ext2add ext2add
+    # Compute the denominator for segment 1 LastRow
+    padw mem_loadw.4294903304 drop drop mem_load.500000101 push.0 ext2sub
+    ext2div # divide the numerator by the divisor
+    exec.compute_boundary_constraints_aux_first
+    # Accumulate the numerator for segment 1 FirstRow
+    ext2add ext2add
+    # Compute the denominator for segment 1 FirstRow
+    padw mem_loadw.4294903304 drop drop push.1 push.0 ext2sub
+    ext2div # divide the numerator by the divisor
+    exec.compute_boundary_constraints_main_first
+    # Accumulate the numerator for segment 0 FirstRow
+    ext2add ext2add
+    # Compute the denominator for segment 0 FirstRow
+    padw mem_loadw.4294903304 drop drop push.1 push.0 ext2sub
+    ext2div # divide the numerator by the divisor
+    # Accumulate point of the constraint polynomial
+    ext2add ext2add
```

Comment:
Denominators for both main and aux segments are the same in the same row (e.g., line 148 is the same as line 154). So, we need to compute denominators for the first and last rows only once.

We can probably precompute denominator, keep it on the stack, and then compute the numerator and perform division (may be faster than saving denominator to memory).

Reply:
> may be faster than saving denominator to memory

The denominator is not saved to memory, the load is for the `z`.

> Denominators for both main and aux segments are the same in the same row ... keep it on the stack

The values are computed using natural order, so:

1. main trace, first row
2. main trace, last row
3. aux trace, first row
4. aux trace, last row

to keep the values is the stack from steps 1/2 to steps 3/4 would require too many operations for it to be worth it (this particular example has very few constraints, for a bigger system it is not a profitable transformation)

Reply:
I meant that the loads in this part of the code are used to load either the value of `z` or the value of `g^{trace_len-2}`, not that it loaded the division of one by the other. Sorry for the confusion.

Reply:
Okay, I went over the code once again, and I was wrong. Only the codegen has to be iterated in the boundary constraint natural order (necessary to compute the correct offsets into the frames when loading data). But once the procedures are emitted the code will be fine, so the call order here doesn't have to match the natural order.

So I can implement what you guys are suggesting. I'm changing it.

Reply:
updated